magento/devdocs#: Expected exceptions (#4148)
magento/devdocs: guides/v2.3/graphql/functional-testing.md

A fixture rollback filename should correspond to the original fixture filename postfixed by `_rollback` keyword. (For example, `virtual_product_rollback.php`.) 
Magento provides fixtures in the `dev/tests/integration/testsuite/Magento/<ModuleName>/_files` directory. Use these fixtures whenever possible. When you create your own fixture, also create a proper rollback.

## Expected exceptions

> Sometimes we need to cover a negative scenarious in our test and make sure that our functionality proper responses on them. There are two ways how you can define the expected exception message:

Reviewer: Please, rephrase the description slightly by removing "we, our etc" occurrences. In case of documentation, such writing style might be a bit confusing. In result, it will be something like: "In cases when it's necessary to cover negative scenarios ... There are two ways how the expected exception message can be defined".

Author: **fixed**

Reviewer 2: Or to minimize further, it could simply be.. "Sometimes you need to cover a negative scenario in your test to ensure..."

It's okay to use "you". Just keep the sentences simple and concise.

Author: Hi, @shrielenee ! Thank you for your notice.

Could you please clarify if this version is ok?

**Current version:** 
```In cases when it's necessary to cover negative scenarios in the test and make sure that testing functionality proper responses on them use the next ways to define the expected exception message:```

Thanks!

Reviewer 2: I don't understand what this phrase means --- "make sure that testing functionality proper responses on them"

I would reword it similarly to...

> When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:

Reviewer 2: But this PR needs a copyedit for language/grammar when the technical content is done.

Author: > I would reword it similarly to...

**implemented**

Thank you, @shrielenee !

> But this PR needs a copyedit for language/grammar when the technical content is done.

@rogyar , do you have any suggestions to the technical content?

Many thanks!

Reviewer 2: Ping me when technical content is done and I'll copyedit.

  - in the body of test function
  - in the test function annotation 

> #### Define exception message in the body of test

Reviewer: Please, replace with "Define exception message expectation ... "

Author: **fixed**

    
    ...
}

```
or 
```php
public function testMyExceptionTest()
{
    ...
    
    $this->expectExceptionMessage("Expected exception message goes here...");
    
    ...
}

```

For example, if `customer A` wants to retrieve information about `customer B cart. In this situation, `customer A` will get an error - "`The current user cannot perform operations on cart "XXXXX`", where `XXXXX` - is an unique ID of `customer B` cart. See code below to understand how to cover this example with `expectExceptionMessage` function:

```php
    /**
     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_items_saved.php
     */
    public function testGetCartFromAnotherCustomer()
    {
        $reservedOrderId = 'test_order_item_with_items';
        $this->quoteResource->load(
            $this->quote,
            $reservedOrderId,
            'reserved_order_id'
        );
        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
        $query = $this->prepareGetCartQuery($maskedQuoteId);
        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
        $this->graphQlQuery($query);
    } 

> ```

Reviewer: I would also add something like:
"Please note that the exception message expectation should be defined before invoking a logic that generates the exception"

Author: **fixed**

@rogyar , could you please review?

#### Define expected exception message in test function annotation

> ```php

Reviewer: Please add a sentence after the header that says: "As an alternative for calling `expectExceptionMessage` method" a predefined directive `@expectedExceptionMessage` can be used for a test

Author: **fixed**

    /**
     * @expectedException \Exception
     * @expectedExceptionMessage Expected exception message goes here...
     */
```

> Let's imagine that customer wants to retrieve information about his own cart but he provides incorrect cart ID.

Reviewer: Please, replace "Let's imagine" with "Example: a customer wants ... "

Author: **fixed**

A fixture rollback filename should correspond to the original fixture filename postfixed by `_rollback` keyword. (For example, `virtual_product_rollback.php`.) 
Magento provides fixtures in the `dev/tests/integration/testsuite/Magento/<ModuleName>/_files` directory. Use these fixtures whenever possible. When you create your own fixture, also create a proper rollback.

## Expected exceptions
When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:
  - in the body of test
  - in test function annotation

> #### Define exception message expectation in the body of test

Reviewer 3: Delete one of the #s. The previous heading was a H2, so this one should be a H3.

Author: **done**

    ...
}
```
{% include note.html type="info" content="Please note that the exception message expectation should be defined before invoking a logic that generates the exception." %}

For example, if `customer A` wants to retrieve information about `customer B cart. In this situation, `customer A` will get an error - "`The current user cannot perform operations on cart "XXXXX`", where `XXXXX` - is an unique ID of `customer B` cart. See code below to understand how to cover this example with `expectExceptionMessage` function:

```php
    /**
     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_items_saved.php
     */
    public function testGetCartFromAnotherCustomer()
    {
        $reservedOrderId = 'test_order_item_with_items';
        $this->quoteResource->load(
            $this->quote,
            $reservedOrderId,
            'reserved_order_id'
        );
        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
        $query = $this->prepareGetCartQuery($maskedQuoteId);
        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
        $this->graphQlQuery($query);
    } 
```

> #### Define expected exception message in test function annotation

Reviewer 3: Change this to a H3 as well

Author: **done**

```php
    /**
     * @expectedException \Exception
     * @expectedExceptionMessage Expected exception message goes here...
     */
```

Example: a customer wants to retrieve information about his own cart but he provides incorrect cart ID.

**Query**
```text
{
  cart(cart_id: "YYYYY") {
    items {
      __typename
      id
      qty
    }
  }
}
```

**Result**
```json
{
  "errors": [
    {
      "message": "Could not find a cart with ID \"YYYYY\"",
      "category": "graphql-no-such-entity",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "cart"
      ]
    }
  ],
  "data": {
    "cart": null
  }
}
```

See the code below how to cover this example with `@expectExceptionMessage` annotation:
```php
    /**
     * @expectedException \Exception
     * @expectedExceptionMessage Could not find a cart with ID "non_existent_masked_id"
     */
    public function testGetNonExistentCart()
    {
        $maskedQuoteId = 'non_existent_masked_id';
        $query = $this->prepareGetCartQuery($maskedQuoteId);
        
        $this->graphQlQuery($query);
    }
```

> There is a list of functions which you can use to cover expected exceptions:  

Reviewer 3: ```suggestion
Use the following functions to cover expected exceptions:  
```

Author: > Suggested change
> There is a list of functions which you can use to cover expected exceptions:  
> Use the following functions to cover expected exceptions:

**done**

> - expectException

Reviewer 3: Wrap the functions with a text code block

Author: > Wrap the functions with a text code block

**done**

Thank you, @keharper 👍

$registry->unregister('isSecureArea');
$registry->register('isSecureArea', false);
```

## Expected exceptions

> When it is necessary to cover negative test scenarios and ensure that testing elicits proper responses, define the expected exception message:

Author: **implemented**

  - in the body of test
  - in test function annotation

> ### Define exception message expectation in the body of test

Author: **fixed**

> 

Author: **done**

Could you please review?

```php
public function testMyExceptionTest()
{
    ...
    
    self::expectExceptionMessage("Expected exception message goes here...");
    
    ...
}

```
or 
```php
public function testMyExceptionTest()
{
    ...
    
    $this->expectExceptionMessage("Expected exception message goes here...");
    
    ...
}
```

{:.bs-callout .bs-callout-info}

> Please note that the exception message expectation should be defined before invoking a logic that generates the exception.

Author: **implemented**

> For example, if `customer A` wants to retrieve information about `customer B cart. In this situation, `customer A` will get an error - "`The current user cannot perform operations on cart "XXXXX`", where `XXXXX` - is an unique ID of `customer B` cart. See code below to understand how to cover this example with `expectExceptionMessage` function:

Author: **implemented**

```php
    /**
     * @magentoApiDataFixture Magento/Checkout/_files/quote_with_items_saved.php
     */
    public function testGetCartFromAnotherCustomer()
    {
        $reservedOrderId = 'test_order_item_with_items';
        $this->quoteResource->load(
            $this->quote,
            $reservedOrderId,
            'reserved_order_id'
        );
        $maskedQuoteId = $this->quoteIdToMaskedId->execute((int)$this->quote->getId());
        $query = $this->prepareGetCartQuery($maskedQuoteId);
        self::expectExceptionMessage("The current user cannot perform operations on cart \"$maskedQuoteId\"");
        $this->graphQlQuery($query);
    } 
```

> ### Define expected exception message in test function annotation

Reviewer 3: ```suggestion
### Exception messages in the annotation of a test function
```

Author: **fixed**

As an alternative for calling `expectExceptionMessage` method use a predefined directive `@expectedExceptionMessage`:

```php
    /**
     * @expectedException \Exception
     * @expectedExceptionMessage Expected exception message goes here...
     */
```

> Example: a customer wants to retrieve information about his own cart but he provides incorrect cart ID.

Reviewer 3: ```suggestion
In the following query, a customer provides an incorrect cart ID while trying to retrieve information about his own cart.
```

Author: **fixed**

**Query**
```text
{
  cart(cart_id: "YYYYY") {
    items {
      __typename
      id
      qty
    }
  }
}
```

**Result**
```json
{
  "errors": [
    {
      "message": "Could not find a cart with ID \"YYYYY\"",
      "category": "graphql-no-such-entity",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "cart"
      ]
    }
  ],
  "data": {
    "cart": null
  }
}
```

> See the code below how to cover this example with `@expectExceptionMessage` annotation:

Reviewer 3: ```suggestion
The `@expectExceptionMessage` annotation provides the text for the exception in this test.
```

Author: **fixed**

Author: Thank you very much for your review , @keharper !
